Implement module-wise testing

Ibaqu opened the pull request:
## Purpose
Modules can now be tested using the `test` command. 

```
ballerina test Module1
```

Fixes #27258

## Approach
Modified it to allow us to directly execute a module in the Modules directory

```
$ballerina test Module1
```

To execute the default module, you simply include the base project name.

```
$ballerina test ProjectA
```

If the module doesn't exist, the following error is immediately printed without compiling. (Error message needs some work. The wording is a bit odd)

```
$ballerina test Module3
error: Cannot execute module Module3. Does not exist in the modules directory
```

The feature works by removing the unnecessary modules from the project and running the project as normal.

Since the default module cannot be removed, extra checks need to be done in the RunTestTask class to purposely skip running the default module. Trying to remove the default module results in a Bad Sad.

If the default module is to be executed, then all the modules are removed and the execution continues. Checks are made to distinguish if the user intends to execute the default module or a module in the modules folder. 

Multiple modules cannot be executed at once, and it results in a `too many arguments` error

## Samples
> Provide high-level details about the samples related to this feature.

## Remarks
> List any other known issues, related PRs, TODO items, or any other notes related to the PR.

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

Dilhasha commented on cli/ballerina-cli/src/main/java/io/ballerina/cli/task/RunTestsTask.java:
```diff
@@ -168,6 +175,18 @@ public void execute(Project project) {
             ModuleName moduleName = module.moduleName();
 
+            if (isModuleTestExecution) {
+                if (moduleExecTests.contains("--all")) {
```
I don't think we are supporting "--all". So, we should be able to get rid of it. WDYT @azinneera ?

Ibaqu replied:
Removed `--all` functionality

Dilhasha commented:
Let's add a test case for module-wise testing.

Dilhasha commented on tests/testerina-integration-test/src/test/resources/project-based-tests/module-execution-tests/tests/main_test.bal:
```diff
@@ -0,0 +1,6 @@
+@test:Config {}
+public function testMain() {
+    test:assertTrue(true, "Not true");
+}
```
missing new line

Dilhasha commented:
Should we support "ballerina test --modules module1,module2" or "ballerina test module1"?
WDYT? @azinneera @hevayo

Ibaqu commented:
I have modified the approach to Module wise testing as follows

1. Removed the `--modules` flag entirely. Multiple modules cannot be executed.
2. Modified it to allow us to directly execute the module

```
ballerina test Module1
```

3. If the module doesn't exist, the following error is immediately printed without compiling. (Error message needs some work. The wording is a bit odd)

```
ballerina test Module3
error: Cannot execute module Module3. Does not exist in the modules directory
```
 
The feature works by removing the modules unnecessary modules from the project and running the project like normal. 

Since the default module cannot be removed, extra checks need to be done in the `RunTestTask` class to purposely skip running the default module. Trying to remove the default module results in a Bad Sad.

Dilhasha commented on cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/TestCommand.java:
```diff
@@ -175,6 +190,39 @@ public void execute() {
                 return;
             }
             isSingleFile = true;
+        } else if (isModule == true) {
```
we can simply use "isModule" instead of "isModule == true"

Ibaqu replied:
Corrected to `isModule`

Dilhasha commented on cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/TestCommand.java:
```diff
@@ -134,15 +140,24 @@ public void execute() {
         } else {
-            args = argList.subList(1, argList.size()).toArray(new String[0]);
-            this.projectPath = Paths.get(argList.get(0));
+            if (ProjectUtils.isBallerinaProject(this.projectPath) &&
```
should we also check whether argList size is only one, if more than 1 we can throw a warning.

Ibaqu replied:
Added a check for arguments greater than one. Added a test case also

Dilhasha commented on tests/testerina-integration-test/src/test/java/org/ballerinalang/testerina/test/ModuleExecutionTest.java:
```diff
@@ -0,0 +1,68 @@
+        clientLeecher2.waitForText(20000);
+    }
+
+    @Test(enabled = false)
```
Why have we disabled this test?

Dilhasha commented on tests/testerina-integration-test/src/test/resources/project-based-tests/module-execution-tests/tests/main_test.bal:
```diff
@@ -0,0 +1,6 @@
+@test:Config {}
+public function testMain() {
+    test:assertTrue(true, "Not true");
+}
```
Missing new line at EOF

Ibaqu replied:
Added nl

Dilhasha commented on cli/ballerina-cli/src/main/java/io/ballerina/cli/task/RunTestsTask.java:
```diff
@@ -197,6 +203,7 @@ public void execute(Project project) {
             if (isSingleTestExecution || isRerunTestExecution) {
                 suite.setTests(TesterinaUtils.getSingleExecutionTests(suite.getTests(), singleExecTests));
             }
+
```
unnecessary new line

Ibaqu replied:
Removed nl

Dilhasha commented on cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/TestCommand.java:
```diff
@@ -134,15 +140,24 @@ public void execute() {
-            args = argList.subList(1, argList.size()).toArray(new String[0]);
-            this.projectPath = Paths.get(argList.get(0));
+            if (ProjectUtils.isBallerinaProject(this.projectPath) &&
+                    !(this.argList.size() > 1) && !ProjectUtils.isBallerinaProject(Paths.get(this.argList.get(0)))) {
```
Let's add "this.argList.size() == 1" instead of "!(this.argList.size() > 1)" ?

Ibaqu commented:
Moving functionality of module based testing to the --tests flag. hence closing
